File: 338-counting-bits/solution.java
import java.util.HashMap;
class Solution {
    HashMap<Integer, Integer> cache;
    
    public Solution() {
        cache = new HashMap<>();
    }
    
    private int countN(int n) {
        if (n == 0) {
            return 0;
        }
        if (n == 1) {
            return 1;
        }
        
        if (cache.get(n) != null) {
            return cache.get(n);
        }
        
        int ret;
        if (n % 2 == 0) {
            ret = countN(n / 2);
        } else {
            ret = countN((n-1) / 2) + 1;
        }
        cache.put(n, ret);
        return ret;
    }
    public int[] countBits(int n) {
        int[] ret = new int[n+1];
        for (int i = 0; i <= n; i++) {
            ret[i] = countN(i);
        }
        
        return ret;
    }
}

File: 338-counting-bits/solution.py
class Solution(object):
    def __init__(self):
        self.cache = {}
        
    def _countN(self, n):
        if n == 0:
            return 0
        if n == 1:
            return 1
        if n in self.cache:
            return self.cache[n]
        
        if n % 2 == 0:
            ret = self._countN(n / 2)
        else:
            ret = self._countN( n / 2) + 1
        self.cache[n] = ret
        return ret
        
    def countBits(self, n):
        """
        :type n: int
        :rtype: List[int]
        """
        ret = []
        for i in range(n+1):
            ret.append(self._countN(i))
            
        return ret
        
        
